src/wasm/wasm-interpreter.cpp

```diff
@@ -19,4 +35,252 @@ void Indenter::print() {
 }
 #endif // WASM_INTERPRETER_DEBUG
 
+//
+// ContextAwareExpressionRunner
+//
+
+const Name ContextAwareExpressionRunner::NONCONSTANT_FLOW =
+  "Binaryen|nonconstant";
+
+ContextAwareExpressionRunner::ContextAwareExpressionRunner(Module* module,
+                                                           Flags flags,
+                                                           Index maxDepth,
+                                                           GetValues* getValues)
+  : ExpressionRunner<ContextAwareExpressionRunner>(maxDepth), module(module),
+    getValues(getValues), flags(flags) {
+  // Trap instead of aborting if we hit an invalid expression.
+  trapIfInvalid = true;
+}
+
+Module* ContextAwareExpressionRunner::getModule() { return module; }
+
+bool ContextAwareExpressionRunner::setLocalValue(Index index,
+                                                 Literals& values) {
+  if (values.isConcrete()) {
+    localValues[index] = values;
+    return true;
+  }
+  localValues.erase(index);
+  return false;
+}
+
+bool ContextAwareExpressionRunner::setLocalValue(Index index,
+                                                 Expression* expr) {
+  auto setFlow = visit(expr);
+  if (!setFlow.breaking()) {
+    return setLocalValue(index, setFlow.values);
+  }
+  return false;
+}
+
+bool ContextAwareExpressionRunner::setGlobalValue(Name name, Literals& values) {
+  if (values.isConcrete()) {
+    globalValues[name] = values;
+    return true;
+  }
+  globalValues.erase(name);
+  return false;
+}
+
+bool ContextAwareExpressionRunner::setGlobalValue(Name name, Expression* expr) {
+  auto setFlow = visit(expr);
+  if (!setFlow.breaking()) {
+    return setGlobalValue(name, setFlow.values);
+  }
+  return false;
+}
+
+Flow ContextAwareExpressionRunner::visitLoop(Loop* curr) {
+  // loops might be infinite, so must be careful
+  // but we can't tell if non-infinite, since we don't have state, so loops
+  // are just impossible to optimize for now
+  return Flow(NONCONSTANT_FLOW);
+}
+
+Flow ContextAwareExpressionRunner::visitCall(Call* curr) {
+  // Traverse into functions using the same mode, which we can also do
+  // when replacing as long as the function does not have any side effects.
+  // Might yield something useful for simple functions like `clamp`, sometimes
+  // even if arguments are only partially constant or not constant at all.
+
+  // Only traverse into functions if not in a function-parallel scenario, where
+  // functions may or may not have been optimized already to something we can
+  // traverse successfully.
+  if (!(flags & FlagValues::PARALLEL)) {
+    // Note that we are not a validator, so we skip calls to functions that do
+    // not exist yet or where the signature does not match.
```

Comment:
This seems a little brittle - the compiler might end up emitting some nondeterministic. Is it worth handling these cases? I lean towards asserting that if we are allowed to traverse other functions, then they exist and are valid.

Reply:
Makes sense, yeah. Change it, and doesn't hit any assertions anymore now that precompute doesn't use it due to parallelism.